File: libs/utils/lib-core/src/main/java/com/yyxnb/what/core/ToastUtils.java
package com.yyxnb.what.core;

import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.Color;
import android.graphics.Typeface;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.ColorInt;
import androidx.annotation.DrawableRes;
import androidx.annotation.NonNull;

import com.yyxnb.what.app.AppUtils;

import java.io.Serializable;

/**
 * 自定义Toast
 *
 * @author yyx
 */
public class ToastUtils implements Serializable {
    private static final @ColorInt
    int DEFAULT_TEXT_COLOR = Color.parseColor("#FFFFFF");

    private static final String TOAST_TYPEFACE = "sans-serif-condensed";

    private static Context context = AppUtils.getApp();

    /**
     * Toast对象
     */
    private static Toast mToast = null;

    private ToastUtils() {
    }

    public static Toast normal(@NonNull String message) {
        return normal(message, Toast.LENGTH_SHORT, null);
    }

    public static Toast normal(@NonNull String message, int duration,
                               Drawable icon) {
        return custom(message, icon, -1, duration);
    }

    public static Toast custom(@NonNull String message, Drawable icon, @DrawableRes int bgRes, int duration) {
        return custom(message, icon, DEFAULT_TEXT_COLOR, bgRes, duration);
    }

    public static Toast custom(@NonNull String message, @DrawableRes int iconRes,
                               @ColorInt int textColor, @DrawableRes int bgRes, int duration) {
        return custom(message, getDrawable(context, iconRes), textColor, bgRes, duration);
    }

    /**
     * 自定义toast方法
     *
     * @param message   提示消息文本
     * @param icon      提示消息的icon,传入null代表不显示
     * @param textColor 提示消息文本颜色
     * @param bgRes     提示背景颜色
     * @param duration  显示时长
     * @return
     */
    public static Toast custom(@NonNull String message, Drawable icon, @ColorInt int textColor, @DrawableRes int bgRes, int duration) {

        if (mToast == null) {
            mToast = new Toast(context);
        } else {
            mToast.cancel();
            mToast = null;
            mToast = new Toast(context);
        }

        @SuppressLint("InflateParams")
        View toastLayout = ((LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE)).inflate(R.layout._toast_layout, null);
        ImageView toastIcon = toastLayout.findViewById(R.id.toast_icon);
        TextView toastText = toastLayout.findViewById(R.id.toast_text);

        if (bgRes != -1) {
            setBackground(toastLayout, getDrawable(context, bgRes));
        }

        if (icon == null) {
            toastIcon.setVisibility(View.GONE);
        } else {
            setBackground(toastIcon, icon);
        }

        toastText.setTextColor(textColor);
        toastText.setText(message);
        toastText.setTypeface(Typeface.create(TOAST_TYPEFACE, Typeface.NORMAL));

        mToast.setView(toastLayout);
        mToast.setDuration(duration);
        mToast.show();
        return mToast;
    }

    @SuppressLint("ObsoleteSdkInt")
    public static void setBackground(View view, Drawable drawable) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
            view.setBackground(drawable);
        } else {
            view.setBackgroundDrawable(drawable);
        }
    }

    public static Drawable getDrawable(Context context, @DrawableRes int id) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            return context.getDrawable(id);
        } else {
            return context.getResources().getDrawable(id);
        }
    }
}

File: widget/lib_common/src/main/java/com/yyxnb/common/utils/ToastUtils.kt
package com.yyxnb.common.utils

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Color
import android.graphics.Typeface
import android.graphics.drawable.Drawable
import android.os.Build
import android.view.LayoutInflater
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.annotation.ColorInt
import androidx.annotation.DrawableRes
import com.yyxnb.common.R
import com.yyxnb.widget.WidgetManager.getContext
import java.io.Serializable

/**
 * 自定义Toast
 *
 * @author yyx
 */
object ToastUtils : Serializable {

    @ColorInt
    private val DEFAULT_TEXT_COLOR = Color.parseColor("#FFFFFF")
    private const val TOAST_TYPEFACE = "sans-serif-condensed"
    private val context = getContext()

    /**
     * Toast对象
     */
    private var mToast: Toast? = null

    fun normal(message: String, duration: Int = Toast.LENGTH_SHORT,
               icon: Drawable? = null): Toast? {
        return custom(message, icon, -1, duration)
    }

    fun custom(message: String, icon: Drawable?, @DrawableRes bgRes: Int, duration: Int): Toast? {
        return custom(message, icon, DEFAULT_TEXT_COLOR, bgRes, duration)
    }

    fun custom(message: String, @DrawableRes iconRes: Int,
               @ColorInt textColor: Int, @DrawableRes bgRes: Int, duration: Int): Toast? {
        return custom(message, getDrawable(context, iconRes), textColor, bgRes, duration)
    }

    /**
     * 自定义toast方法
     *
     * @param message   提示消息文本
     * @param icon      提示消息的icon,传入null代表不显示
     * @param textColor 提示消息文本颜色
     * @param bgRes     提示背景颜色
     * @param duration  显示时长
     * @return
     */
    fun custom(message: String, icon: Drawable?, @ColorInt textColor: Int, @DrawableRes bgRes: Int, duration: Int): Toast? {
        if (mToast == null) {
            mToast = Toast(context)
        } else {
            mToast!!.cancel()
            mToast = null
            mToast = Toast(context)
        }
        @SuppressLint("InflateParams") val toastLayout = (context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater).inflate(R.layout._toast_layout, null)
        val toastIcon = toastLayout.findViewById<ImageView>(R.id.toast_icon)
        val toastText = toastLayout.findViewById<TextView>(R.id.toast_text)
        if (bgRes != -1) {
            setBackground(toastLayout, getDrawable(context, bgRes))
        }
        if (icon == null) {
            toastIcon.visibility = View.GONE
        } else {
            setBackground(toastIcon, icon)
        }
        toastText.setTextColor(textColor)
        toastText.text = message
        toastText.typeface = Typeface.create(TOAST_TYPEFACE, Typeface.NORMAL)
        mToast!!.view = toastLayout
        mToast!!.duration = duration
        mToast!!.show()
        return mToast
    }

    @SuppressLint("ObsoleteSdkInt")
    fun setBackground(view: View, drawable: Drawable?) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
            view.background = drawable
        } else {
            view.setBackgroundDrawable(drawable)
        }
    }

    fun getDrawable(context: Context, @DrawableRes id: Int): Drawable? {
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            context.getDrawable(id)
        } else {
            context.resources.getDrawable(id)
        }
    }
}